### ISSUE TITLE
[apex] Catch Specific Exception Type

### ISSUE THREAD

**Issue by mitchspano** [pmd/pmd#5426]
**Proposed Rule Name:** [apex] Catch Specific Exception Type

**Proposed Category:** Best Practices

**Description:** This rule enforces the best practice of catching specific exceptions before catching the generic `Exception` class in Apex try/catch blocks. Catching the generic `Exception` can mask underlying issues and make debugging more difficult. Developers should catch the specific exception type thrown by the method call.

**Code Sample:**

```java
// Violates the rule: Only generic Exception is caught
try {
    // Some code that could throw exceptions
} catch (Exception e) {
    // Generic exception handling
}

// Compliant code: Specific exceptions are caught first
try {
    // Some code that could throw exceptions
} catch (DmlException e) {
    // Handle DML exception
} catch (NullPointerException e) {
    // Handle Null Pointer exception
} catch (Exception e) {
    // Handle any other exception
}
```

**Possible Properties:**

* **IgnoreTestClasses**: (Boolean):  A flag to indicate whether the rule should be ignored in test classes. This can be useful as test classes sometimes require more lenient exception handling.

**Comment by sirephil** [comment]
Doing:

```
} catch (Exception e) {
} catch (SomeException e) {
```

Should be invalid (I never tested to see if the Apex compiler complains). If it is allowed, I totally agree with having this rule.

**Comment by pkozuchowski** [comment]
Description sounds good, example does not. Just catching general exception shouldn't be invalid. Catching all possibile exceptions means that your top-level class needs to know all implementation details of everything on lower level rather than treating it as a black box. What's worse, it also creates dependency to those exceptions.
This sounds like violation of Dependency inversion principle, because it's enforcing you to work on concrete implementations  (SomeException) rather than an interface (Exception), when you don't need to.

Consider following example:

- Transaction starts with AuraHandled method, which calls some high level business logic. All this method cares about is that if there's an exception, it should be logged and rethrown as AuraHandledException. Otherwise, just delegate work to business logic service
- Business Logic may be composed of smaller blocks from different packages. If I were to catch all exceptions, then:
- 1) There's no benefit from doing so for me. It just reduces readability and increases code footprint.
- 2) I need to figure out what all the black boxes do and throw
- 3) There's a compiled dependency with a concrete exception class, which complicates refactoring.

I've just checked and compiler does not allow you to catch anything after Exception:
```apex
try {
    throw new DmlException('test');
} catch (Exception e) {
} catch (DmlException e) {

}

Compile failure on line 4, column 26: Exception type already caught: System.DmlException

```

**Comment by mitchspano** [comment]
Thanks for the feedback @pkozuchowski,  I've updated the example to be more illustrative of what we are trying to go for here.

The point you made about a compile-time dependency on the specific exception type is definitely valid. This rule proposal really originates from best practices for Java, where the method signature indicates which exception types it can throw using the `throws` keyword, so there's definitely something lost in translation between Java to Apex.

I agree that sometimes it is necessary to catch `Exception`, like when calling a method from a package. In my experience, it is more often the case that developers will just carelessly use the catch all `Exception` and neglect to properly handle the error.

Perhaps we should relax this rule specification such that if the try block contains a method call from a class with a namespace, then generic `Exception` is permitted and no PMD violation will be reported.
Are there any other instances which we would be able to deterministically define a scenario where specific exception types cannot be determined?

For the record, I never said we should do something like this:
```java
} catch (Exception e) {
} catch (SomeException e) {
```
I'm not sure where that came from, but that's just silly and doesn't compile.

**Comment by pkozuchowski** [comment]
To be honest, there is too few valid use cases for this exception handling to justify enforcing it as a standard. Other than flagging all catch(Exception), I don't see how this rule could determine which exceptions should be caught. It would generate a lot of false positives without adding any real value. In 99% of cases, you won't need different behavior for different exceptions.

**Comment by DougMidgley** [comment]
> To be honest, there is too few valid use cases for this exception handling to justify enforcing it as a standard. Other than flagging all catch(Exception), I don't see how this rule could determine which exceptions should be caught. It would generate a lot of false positives without adding any real value. In 99% of cases, you won't need different behavior for different exceptions.

Second this comment

**Comment by sirephil** [comment]
> > To be honest, there is too few valid use cases for this exception handling to justify enforcing it as a standard. Other than flagging all catch(Exception), I don't see how this rule could determine which exceptions should be caught. It would generate a lot of false positives without adding any real value. In 99% of cases, you won't need different behavior for different exceptions.
> 
> Second this comment

Ditto.
